File: patterns/linked-list-reversal/03.reverse-nodes-in-k-groups/main.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        group_prev = dummy

        while True:
            kth = group_prev
            count = k
            while kth and count > 0:
                kth = kth.next
                count -= 1

            if not kth:
                break
            
            group_next = kth.next
            prev, curr = kth.next, group_prev.next

            while curr != group_next:
                next = curr.next
                curr.next = prev
                prev = curr
                curr = next
            
            tmp = group_prev.next
            group_prev.next = kth
            group_prev = tmp

        return dummy.next

File: patterns/linked-list-reversal/03.reverse-nodes-in-k-groups/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode dummy(0, head);
        ListNode* group_prev = &dummy;
        
        while (true) {
            ListNode* kth = group_prev;
            int count = k;
            while (kth && (count > 0)) {
                kth = kth->next;
                count--;
            }

            if (kth == nullptr) break;

            ListNode* group_next = kth->next;
            ListNode* prev = kth->next;
            ListNode* curr = group_prev->next;

            while(curr != group_next) {
                ListNode* next = curr->next;
                curr->next = prev;
                prev = curr;
                curr = next;
            }

            ListNode* tmp = group_prev->next;
            group_prev->next = kth;
            group_prev = tmp;
        }

        return dummy.next;
    }
};
